### PR TITLE
graphql,metadata,db: support queries for tables in all schemas

### LINKED ISSUE #3: graphql, metadata: support for querying tables that are not in the "public" schema
- [x] #11
- [x] #10
- [x] #9

### LINKED ISSUE #9: Add support to `export` and `import` metadata

### LINKED ISSUE #10: Add ability to search through the tracked tables to get more info on the right schema in SQL gen step

### LINKED ISSUE #11: Add support for `metadata`
- [x] Support for the ability to `track` and `untrack` tables

### PR DESCRIPTION
**Pull request by kolharsam**
The PR hopes to 

resolve #3, resolve #9, resolve #10 and resolve #11

### CODE REVIEW

File: src/server.rs [hunk 1]
```diff
@@ -24,16 +25,35 @@ pub async fn healthz_handler(
 pub enum MetadataMessage {
     TrackTable(QualifiedTable),
     UntrackTable(QualifiedTable),
-    // NOTE: args will only be `null` in this case
+    // NOTE: args will be `null` for `export_metadata`
     ExportMetadata,
+    ImportMetadata(Metadata),
 }
```
Comment (kolharsam) [inline src/server.rs:31]: These 4 operations on metadata are supported.

- `track_table`
- `untrack_table`
- `export_metadata`
- `import_metadata`

**Comment by iykekings** [comment]
Code review almost done

File: src/server.rs [hunk 2]
```diff
@@ -24,16 +25,35 @@ pub async fn healthz_handler(
...
+    actix_web::web::HttpResponse::build(status).json(ErrorResponse {
+        error: String::from(err_message),
+    })
+}
+
+fn bad_request_response(err_message: &str) -> actix_web::HttpResponse {
+    // NOTE: probably should be making use of currying here
+    make_metadata_error_response(StatusCode::BAD_REQUEST, err_message)
+}
+
```
Comment (iykekings) [inline src/server.rs:56]: We can follow in the footsteps of `MetadataMessage`, use an enum to represent the Metadata response and impl Responder for it. So we can return it in any function that returns impl Responder like. So we will replace this
```suggestion
#[derive(Serialize, Debug, Clone)]
pub enum MetadataResponse {
    Success(String),
    Data(Metadata),
    Error(String),
}

impl Responder for MetadataResponse {
    type Error = actix_web::Error;
    type Future = actix_web::HttpResponse;

    fn respond_to(self, _: &actix_web::HttpRequest) -> Self::Future {
        match self {
            MetadataResponse::Error(err) => actix_web::HttpResponse::build(StatusCode::BAD_REQUEST)
                .json(json!({ "error": err })),
            MetadataResponse::Success(msg) => {
                actix_web::HttpResponse::Ok().json(json!({"success": true, "message": msg}))
            }
            MetadataResponse::Data(metadata) => actix_web::web::HttpResponse::Ok().json(metadata),
        }
    }
}
```

File: src/server.rs [hunk 3]
```diff
@@ -51,35 +71,40 @@ impl DataResponse {
...
-    let parse_result = match payload_to_str {
-        Ok(r) => json::parse(r),
-        Err(_err) => Err(json::Error::FailedUtf8Parsing),
-    };
-
-    let body = parse_result.unwrap_or_else(|e| json::object! { "error" => e.to_string() });
+pub async fn metadata_handler(
+    app_state: actix_web::web::Data<context::AppState>,
+    payload: actix_web::web::Json<MetadataMessage>,
+) -> actix_web::HttpResponse {
```
Comment (iykekings) [inline src/server.rs:77]: so we return `impl Responder` here
```suggestion
) -> impl Responder {
```

File: src/server.rs [hunk 4]
```diff
@@ -51,35 +71,40 @@ impl DataResponse {
...
-                error: e.to_string(),
-            },
-        ),
+    match payload.into_inner() {
+        MetadataMessage::TrackTable(table) => {
+            match (*server_ctx).metadata_track_table(table.clone()) {
+                Ok(_) => make_metadata_success_response(
+                    format!("{} is now being tracked!", table.to_string()).as_str(),
+                ),
+                Err(err) => bad_request_response(format!("{}", err).as_str()),
```
Comment (iykekings) [inline src/server.rs:86]:
```suggestion
              Ok(_) => MetadataResponse::Success(format!(
                    "{} is now being tracked!",
                    table.to_string()
                )),
                Err(err) => MetadataResponse::Error(err.to_string()),
```

File: src/server.rs [hunk 5]
```diff
@@ -51,35 +71,40 @@ impl DataResponse {
...
+            }
+        }
+        MetadataMessage::UntrackTable(table) => {
+            match (*server_ctx).metadata_untrack_table(table.clone()) {
+                Ok(_) => make_metadata_success_response(
+                    format!("{} has now been un-tracked!", table.to_string()).as_str(),
+                ),
+                Err(err) => bad_request_response(format!("{}", err).as_str()),
```
Comment (iykekings) [inline src/server.rs:94]:
```suggestion
                Ok(_) => MetadataResponse::Success(format!(
                    "{} has now been un-tracked!",
                    table.to_string()
                )),
                Err(err) => MetadataResponse::Error(err.to_string()),
```

File: src/server.rs [hunk 6]
```diff
@@ -51,35 +71,40 @@ impl DataResponse {
...
+            }
+        }
+        MetadataMessage::ExportMetadata => {
+            actix_web::web::HttpResponse::Ok().json((*server_ctx).get_metadata())
```
Comment (iykekings) [inline src/server.rs:98]: Shamelessly adding clone here 😅
```suggestion
          MetadataResponse::Data((*server_ctx).get_metadata().clone())
```

File: src/server.rs [hunk 7]
```diff
@@ -51,35 +71,40 @@ impl DataResponse {
...
+        }
+        MetadataMessage::ImportMetadata(md) => {
+            (*server_ctx).replace_metadata(&md);
+            make_metadata_success_response("Imported metadata successfully!")
```
Comment (iykekings) [inline src/server.rs:102]:
```suggestion
           MetadataResponse::Success("Imported metadata successfully!".to_string())
```

**Review by iykekings — CHANGES_REQUESTED** [review]
@kolharsam I have some refactoring suggestion. It is totally optional to adopt it. I will test the functionality in a bit

**Comment by iykekings** [comment]
The `Responder` we impl for `MetadataResponse` defaults to `StatusCode::BadRequest`, but we can return `(MetadataResponse, StatusCode)` where we need to add a custom status code, since `Responder` is implemented for `(T: Responder, StatusCode)`

**Comment by kolharsam** [comment]
@iykekings These are excellent suggestions! :raised_hands: 
Thank you very much! For telling me about this!

File: src/server.rs (hunk 6 above)
Comment (kolharsam) [reply src/server.rs:98]: I don't think clone is ever a bad idea. We can always ask for help from the `--release` flag! :laughing:

**Comment by kolharsam** [comment]
@iykekings I'll be creating an issue out of this recent suggestion(s) that you've made. To implement the `Responder` trait for all the other types that we're making responses from throughout the codebase!

**Comment by kolharsam** [comment]
@iykekings, could you please approve this PR if there are no other objections from your end?

**Review by iykekings — APPROVED** [review]
LGTM! 😎💯

**Pull request merged**